### PR TITLE
Wait for reassemble on hot reload

### LINKED ISSUE #26503: Reassemble method called too late
Consider a widget that rebuilds very often (typically because of an animation). 
Then consider editing the sources of that widgets through hot-reload. 

It seems that when doing so, there's a chance for `build` method to be called _before_ `reassemble` was, but with the updated content. 

This is a big issue for the following open source library: https://github.com/rrousselGit/flutter_hooks
As it needs to override the `reassemble` behavior to handle potential changes within `build` method.

But if `build` method is called before `reassemble`, then the library won't be ready for it and the app will enter into an invalid state.

### PR DESCRIPTION
**Pull request by dnfield**
## Description

In the documentation for reassemble, we claim that we will call it before calling build again on hot reload.  Unfortunately, this doesn't always happen when animations are at play, and is a problem for clients like Flutter Hooks that rely on `reassemble` being called before `build`.

This patch does not solve the problem in a universal way - it is still possible that someone could stuff business logic in a `Timer` and expect that we effectively pause the timer.  This patch would allow for solving that by overriding the `waitForReassemble` method that we added - solving the general case here though would likely require changes to the Dart VM that are very non-trivial and potentially risky, and it would be best to have a concrete use case before attempting them.

This patch does cause animations to pause for the duration of the hot reload, and then start at their new point in time - i.e., it drops the frames that would have rendered during a hot reload.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/26503
A previous attempt was discussed at https://github.com/flutter/flutter/pull/26983.  Unlike that attempt, this one does not hijack the system resume status, and does not cause the engine to stop animating (we just stop listening to the engine's animation callbacks).

## Tests

I added the following tests:

Binding tests, test to make sure we restore the callbacks when we call reassembleApplication.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/foundation/binding.dart [hunk 1]
```diff
@@ -231,6 +239,27 @@ abstract class BindingBase {
...
+  /// Cause the entire application to stop drawing, e.g. in preparation for hot
+  /// reload.
+  ///
+  /// This is used by development tools when the application code is being
+  /// changed, to ensure that animations will not continue to allow builds to
+  /// fire before `ext.flutter.reassemble` is sent. It can be triggered
+  /// manually by sending the `ext.flutter.waitForReassemble` service extension
+  /// signal.
+  ///
+  /// This method must never be used in production code, as it will prevent all
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/foundation/binding.dart]: If this should never be called, should the body be wrapped in an assert?

File: packages/flutter/lib/src/foundation/binding.dart [hunk 2]
```diff
@@ -231,6 +239,27 @@ abstract class BindingBase {
...
+  /// drawing until [reassembleApplication] is called, which is also not meant
+  /// for production code. It is only meant for use in hot reload.
+  @mustCallSuper
+  @visibleForTesting
+  Future<void> waitForReassemble() async {
+    _lastOnBeginFrame ??= window.onBeginFrame;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/foundation/binding.dart]: Name these with debug?

File: packages/flutter/lib/src/foundation/binding.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/binding.dart]: That sounds reasonable to me.  It does make me wonder why `reassembleApplication` is not wrapped in an assert, but that shouldn't stop us from ding it here.

File: packages/flutter/lib/src/foundation/binding.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/binding.dart]: Done

File: packages/flutter/lib/src/foundation/binding.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/binding.dart]: Done

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 3]
```diff
@@ -596,179 +596,195 @@ class HotRunner extends ResidentRunner {
...
-
-    final Stopwatch devFSTimer = Stopwatch()..start();
-    final UpdateFSReport updatedDevFS = await _updateDevFS();
-    // Record time it took to synchronize to DevFS.
-    _addBenchmarkData('hotReloadDevFSSyncMilliseconds', devFSTimer.elapsed.inMilliseconds);
-    if (!updatedDevFS.success)
-      return OperationResult(1, 'DevFS synchronization failed');
-    String reloadMessage;
-    final Stopwatch vmReloadTimer = Stopwatch()..start();
+    final Stopwatch reloadTimer = Stopwatch();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/run_hot.dart]: This is a big diff. Given that our current testing story is inadequate, how confident do you feel about this change?

Comment (dnfield) [reply packages/flutter_tools/lib/src/run_hot.dart]: If you view it without whitespace it's a bit smaller (https://github.com/flutter/flutter/pull/32514/files?utf8=✓&diff=split&w=1).  

The risks in this patch are:

- I messed up some timing/benchmarking. That should show up in the benchmarks - this is not expected to cause any regressions on that.
- I messed up some logic and hot reload could result in causing the application to never draw again.  I think the tests I added should help make sure that doesn't happen, but they don't cover any integration scenario of testing this logic in run_hot.dart. However, I think our benchmarks for hot reload would start failing if it really was that broken.

There's definitely potential for some edge case being missed here. If you can think of another test (maybe another integration test) that covers something we don't have that helps this I'd be happy to add it to this PR.

**Comment by jonahwilliams** [comment]
> This patch does cause animations to pause for the duration of the hot reload, and then start at their new point in time - i.e., it drops the frames that would have rendered during a hot reload.

Is this a desirable behavior, or should we be pausing these animations in a way that also stops the scheduler? 

Hijacking the system channel is not the right approach, but this still feels like a hack

**Comment by dnfield** [comment]
We could pump this to the engine side and tell the animator to stop animating.

**Comment by jonahwilliams** [comment]
We also need some sort of test, devicelab most likely, that creates a scenario which would otherwise cause an error if reassemble is called after build that is then fixed with this change

File: packages/flutter/lib/src/foundation/binding.dart [hunk 4]
```diff
@@ -267,6 +299,23 @@ abstract class BindingBase {
   @protected
   Future<void> performReassemble() {
     FlutterError.resetErrorCount();
+    assert(() {
+      final bool onBeginWasNull = window.onBeginFrame == null;
+      if (_debugLastOnBeginFrame != null) {
+        assert(window.onBeginFrame == null);
+        window.onBeginFrame = _debugLastOnBeginFrame;
+      }
+      assert( onBeginWasNull || window.onBeginFrame != null);
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/binding.dart]: nit: remove extra whitespace

File: packages/flutter/lib/src/foundation/binding.dart [hunk 5]
```diff
@@ -231,6 +236,33 @@ abstract class BindingBase {
...
+  @mustCallSuper
+  @visibleForTesting
+  Future<void> waitForReassemble() async {
+    assert(() {
+      _debugLastOnBeginFrame ??= window.onBeginFrame;
+      _debugLastOnDrawFrame ??= window.onDrawFrame;
+      window.onDrawFrame = null;
+      window.onBeginFrame = null;
+      return true;
+    }());
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/binding.dart]: this is at the wrong layer. The foundation binding doesn't know about these IIRC.

**Comment by Hixie** [comment]
This isn't really a solution to the problem, I don't think we should go down this path.

Specifically, consider a timer that keeps calling a method on an object that is going to do something during reassemble. Nothing stops the method from being called between the hot reload and the reassemble.

**Comment by Hixie** [comment]
The only real solution is to actually run code synchronously after hot reload. We've seen two proposals for doing this:
* have the VM itself call an endpoint when the hot reload is done. If we did this, we could furthermore have that endpoint be given a list of all the objects that were modified or some such, which we could use to e.g. make hot reload faster (only marking dirty objects dirty), etc.
* do the hot reload by setting a breakpoint on a method that we call before the hot reload, waiting for the breakpoint to be hit, doing the hot reload, then resuming execution.

**Comment by dnfield** [comment]
@Hixie - I had spent some time investigating those approaches and don't think they're viable.  Setting a breakpoint would, as far as I can tell, be the same as pausing the Isolate, which doesn't work because we need the isolate to continue running so we can reload sources.

Having the VM itself call an endpoint when it's done wouldn't stop the timers from running in your scenario.

I'm not convinced we can sensibly solve the general scenario of a timer doing business logic.  This fix is meant to be geared only towards fixing animations and frame pumps as the cause of this error.

File: packages/flutter/lib/src/foundation/binding.dart (hunk 5 above)
Comment (dnfield) [reply packages/flutter/lib/src/foundation/binding.dart]: Window was introduced into foundation a while back now.  Are you suggesting that needs to be backed out?  I don't disagree, but it's been in there for a while.

**Comment by dnfield** [comment]
If this approach doesn't work, I'm inclined to say we should just change the documentation on reassemble to clarify cases where it may not be the first thing called after hot reload.

**Comment by dnfield** [comment]
Alternatively, perhaps we could just send a service extension method warning that hot reload is about to start, and let implementers decide what they need to stop if it matters to them.

**Comment by Hixie** [comment]
> Setting a breakpoint would, as far as I can tell, be the same as pausing the Isolate, which doesn't work because we need the isolate to continue running so we can reload sources.

I'm pretty sure you can hot reload while a breakpoint is set. We have tests that check that it's possible (in the flutter_tools integration tests directory). I had to fix bugs to make the tests work, so I'm pretty sure this does in fact work.

> Having the VM itself call an endpoint when it's done wouldn't stop the timers from running in your scenario.

Why not? If the VM calls the reassemble endpoint synchronously from the end of the hot reload, how can the timers run between the reload and the reassemble?

**Comment by goderbauer** [comment]
Closing this as @dnfield will try a different approach.

**Pull request closed without merging**